src/persist-client/src/cli/inspect.rs

```diff
@@ -630,6 +643,204 @@ pub async fn unreferenced_blobs(args: &StateArgs) -> Result<impl serde::Serializ
     Ok(unreferenced_blobs)
 }
 
+/// Roundtrips all blobs through arrow-rs and ensures they match what is currently in S3.
+pub async fn arrow_rs_roundtrip(blob_uri: &str) -> Result<impl Serialize, anyhow::Error> {
+    let cfg = PersistConfig::new_default_configs(&READ_ALL_BUILD_INFO, SYSTEM_TIME.clone());
+
+    let metrics_registry = MetricsRegistry::new();
+    let metrics = Arc::new(Metrics::new(&cfg, &metrics_registry));
+
+    #[derive(Debug, Default, Serialize)]
+    struct BlobReport {
+        num_shards: usize,
+        num_blobs: usize,
+        missing_blobs: usize,
+        num_rows: usize,
+        num_bytes: usize,
+    }
+
+    /// Helper method that lists all shards in a Blob store.
+    async fn list_shards(
+        blob: Arc<dyn Blob + Send + Sync>,
+    ) -> Result<BTreeMap<ShardId, Vec<PartialBatchKey>>, anyhow::Error> {
+        let mut shards = BTreeMap::new();
+        blob.list_keys_and_metadata(&BlobKeyPrefix::All.to_string(), &mut |metadata| {
+            match BlobKey::parse_ids(metadata.key) {
+                Ok((shard, PartialBlobKey::Batch(writer_key, part_id))) => {
+                    let partial_keys = shards.entry(shard).or_insert_with(Vec::new);
+                    partial_keys.push(PartialBatchKey::new(&writer_key, &part_id));
+                }
+                Err(err) => {
+                    eprintln!("error parsing blob: {}", err);
+                }
+                _ => (),
+            }
+        })
+        .await?;
+
+        Ok(shards)
+    }
+
+    // Open a handle to S3.
+    let blob = make_blob(&cfg, blob_uri, NO_COMMIT, Arc::clone(&metrics))
+        .await
+        .context("making blob")?;
+    tracing::info!(?blob_uri, "opened blob");
+
+    // List all shards.
+    let shards = mz_ore::retry::Retry::default()
+        .max_tries(5)
+        .retry_async(|_| {
+            let blob = Arc::clone(&blob);
+            async move { list_shards(Arc::clone(&blob)).await }
+        })
+        .await?;
+
+    let num_parts: usize = shards.values().map(|parts| parts.len()).sum();
+    tracing::info!(
+        ?blob_uri,
+        num_shards = shards.len(),
+        num_parts,
+        "listed blob"
+    );
+
+    let mut report = BlobReport::default();
+    for (shard_id, partial_keys) in shards {
+        // Drives all of the GET futures concurrently.
+        //
+        // Note(parkmycar): We should limit the number of blobs we fetch in parallel to reduce
```

Comment:
Feels like we should probably do this before merging? Otherwise, I have to imagine this won't work for many interesting prod envs.

Reply:
Yeah good call, changed this code to use a bounded channel which because we explicitly acquire a permit before fetching, should bound the number of inflight blobs we have